content/en/blog/_posts/2022-04-19-non-graceful-node-shutdown.md

```diff
@@ -0,0 +1,63 @@
+---
+layout: blog
+title: "Kubernetes 1.24: Introducing Non-Graceful Node Shutdown Alpha"
+date: 2022-04-19T10:00:00-08:00
+slug: kubernetes-1-24-non-graceful-node-shutdown-alpha
+---
+
+**Authors** Xing Yang and Yassine Tijani (VMware)
+
+Kubernetes v1.24 introduces alpha support for [Non-Graceful Node Shutdown](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2268-non-graceful-shutdown). This feature allows stateful workloads to failover to a different node after the original node is shutdown or in a non-recoverable state such as the hardware failure or broken OS.
+
+## How is this different from Graceful Node Shutdown
+
+You might have heard about the [Graceful Node Shutdown](https://kubernetes.io/blog/2021/04/21/graceful-node-shutdown-beta/) feature and are wondering how the Non-Graceful Node Shutdown feature is different from that. Graceful Node Shutdown introduced a way to detect a node shutdown and handle it gracefully. A Node Shutdown can be `Graceful` only if the node shutdown action can be detected by Kubelet's Node Shutdown Manager. However, there are cases where a node shutdown action may not be detected by Kubelet's Node Shutdown Manager. This could happen either because the shutdown command does not trigger the inhibitor locks mechanism used by Kubelet or because of a user error, i.e., the `ShutdownGracePeriod` and `ShutdownGracePeriodCriticalPods` are not configured properly.
+
+When a node is shutdown but not detected by Kubelet's Node Shutdown Manager, the pods that are part of a StatefulSet will be stuck in terminating status on the shutdown node and cannot move to a new running node. As a result, the application running on the StatefulSet cannot function properly. If the original shutdown node comes up, the pods will be deleted by Kubelet and new pods will be created on a different running node. If the original shutdown node does not come up, these pods will be stuck in terminating status on the shutdown node forever.
```

Comment:
```suggestion
When a node is shutdown but but without the kubelet having an opportunity to detect it, Pods on that node
also shut down ungracefully. For stateless apps, that's often not a problem (a ReplicaSet adds a new Pod once
the cluster detects that the affected node or Pod has failed). For stateful apps, the story is more complicated.
If you use a StatefulSet and have a Pod from that StatefulSet on a node that fails uncleanly, that affected Pod
will be marked as terminating; the StatefulSet cannot create a replacement Pod because the existing Pod
still exists in the API.
As a result, the application running on the StatefulSet may be degraded or even offline. If the original, shut
down node comes up again, the kubelet on that original node reports in, deletes the existing Pods, and
the control plane makes a replacement Pod for that StatefulSet (possible on a different node).
If the original node has failed and does not rejoin your cluster, those stateful pods would be stuck ina
terminating status on that failed node indefinitely.
```